### ISSUE TITLE
handling <sos> and <eos> 

### ISSUE THREAD

**Issue by riyijiye** [vrenkens/nabu#29]
Hi Vincent,

my understanding of using <sos> and <eos> in encoder-decoder network is as follows

say the case with input f1, f2, f3 and target p1, p2 (f1, f2, f3 -> p1, p2), suppose input is padded with max_length 5, and target is padded with max_length 4

1)encoder input: f1, f2, f3, EOS, PAD
2)decoder input: SOS, p1, p2, PAD
3)target: p1, p2, EOS, PAD

from what I saw you did 2) in your rnn_decoder.py code, and 3) in your loss_functions.py (cross_entropy_eos) in your trainer directory.  I guess in terms of speech feature input, 0 can be used as encoder input <eos>, so first padding zero is automatically used? then what if some utterances in the batch actually has max_length and no zero padded?

second question, I am wondering how important it is to separate the loss computation of <eos> from the regular output targets?

thanks,
Harry

**Comment by riyijiye** [comment]
actually for my first question, usually there must be some silences in the original speech so I guess whether padding zero exists or not does not matter, right?

**Comment by vrenkens** [comment]
Hi Harry,

first question: The padding is not actually used, you give the sequence length to the rnn so it will ignore the padding.

I don't quite understand your second question. What do you mean with separating the loss computation from the regular output targets?

**Comment by riyijiye** [comment]
Hi Vincent，

for the first question, I am thinking why you did not add EOS at end of encoder input, as in 1).
for the second question, sorry some miss spelling, we see from below your code EOS CE loss computation is separated.  

#compute the loss for the loss for the sequences
                stacked_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
                    logits=stacked_logits,
                    labels=stacked_targets)

                #compute the loss for the eos label
                eos_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
                    logits=final_logits,
                    labels=final_targets)

**Comment by vrenkens** [comment]
An EOS label in the encoder input is not needed. The label is needed in the decoder because at decode time the model does not know how long the sequence is. It thus needs to be trained to emit an EOS label to stop outputting labels. The length of the input is known at decode time so no EOS label is required.

The EOS loss is computed separately because the EOS labels are not present in  the targets, so they need to be added unlike the actual targets.

**Comment by riyijiye** [comment]
thanks Vincent. I actually meant attaching EOS to the targets and do one CE loss computation. But I think there should not be much difference.
